src/compiler/reflection.jl

```diff
@@ -28,15 +28,42 @@ function reflect(@nospecialize(func), @nospecialize(A), @nospecialize(types);
     return llvmf, mod
 end
 
+# For VERSION >= v"1.9.0-DEV.516"
+struct jl_llvmf_dump
+    TSM::LLVM.API.LLVMOrcThreadSafeModuleRef
+    F::LLVM.API.LLVMValueRef
+end
+
 function enzyme_code_llvm(io::IO, @nospecialize(func), @nospecialize(A), @nospecialize(types);
                           optimize::Bool=true, run_enzyme::Bool=true, second_stage::Bool=true,
                           raw::Bool=false, debuginfo::Symbol=:default, dump_module::Bool=false)
     JuliaContext() do ctx
-        llvmf, _ = reflect(func, A, types; optimize, run_enzyme, second_stage, ctx)
-
-        str = ccall(:jl_dump_function_ir, Ref{String},
-                    (LLVM.API.LLVMValueRef, Bool, Bool, Ptr{UInt8}),
-                    llvmf, !raw, dump_module, debuginfo)
+        entry_fn, ir = reflect(func, A, types; optimize, run_enzyme, second_stage, ctx)
+        @static if VERSION >= v"1.9.0-DEV.516"
+            ts_mod = ThreadSafeModule(ir; ctx)
+            if VERSION >= v"1.9.0-DEV.672"
+                GC.@preserve ts_mod entry_fn begin
+                    value = Ref(jl_llvmf_dump(ts_mod.ref, entry_fn.ref))
+                    str = ccall(:jl_dump_function_ir, Ref{String},
+                          (Ptr{jl_llvmf_dump}, Bool, Bool, Ptr{UInt8}),
+                          value, !raw, dump_module, debuginfo)
+                end
+            else
+                GC.@preserve ts_mod entry_fn begin
+                    # N.B. jl_dump_function_ir will `Libc.free` the passed-in pointer
+                    value_ptr = reinterpret(Ptr{jl_llvmf_dump},
+                                            Libc.malloc(sizeof(jl_llvmf_dump)))
```

Comment:
Do we not need to free this later? Alternatively, could a stack allocation not suffice?

Reply:
Unfortunately not, it's a consequence of the underlying function deleting the pointer in DEV.516. This was rectified in DEV.672, but compatibility in between is handled by this block.